test/test_torch.py

```diff
@@ -16330,16 +16329,140 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+    'ceil': ufunc_meta(is_floating=True),
+    'expm1': ufunc_meta(is_floating=True, complex_on=()),
+    'floor': ufunc_meta(is_floating=True),
+    'log': ufunc_meta(is_floating=True),
+    'log10': ufunc_meta(is_floating=True),
+    'log1p': ufunc_meta(is_floating=True, complex_on=()),
+    'log2': ufunc_meta(is_floating=True),
+    'sin': ufunc_meta(is_floating=True),
+    'sinh': ufunc_meta(is_floating=True),
+    'sqrt': ufunc_meta(is_floating=True),
+    'trunc': ufunc_meta(is_floating=True),
+    'digamma': ufunc_meta(is_floating=True, complex_on=()),
+    'rsqrt': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'sigmoid': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'atan': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'cos': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'cosh': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'exp': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'tan': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'tanh': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'erf': ufunc_meta(is_floating=True, complex_on=()),
+    'erfc': ufunc_meta(is_floating=True, complex_on=()),
+    'erfinv': ufunc_meta(is_floating=True, complex_on=()),
+    'lgamma': ufunc_meta(is_floating=True, complex_on=())
+}
+
+# Creates and adds a test of the given unary floating ufunc's type promotion
+# Unary floating ufuncs promote tensors of bool and integral types to
+# the default scalar type.
+def generate_unary_floating_ufunc_promo_test(cls, op_str):
+    int_types = [torch.bool, torch.uint8, torch.int8, torch.int16,
+                 torch.int, torch.long]
+    float_types = [torch.float, torch.double]
+    float_types_cuda = [torch.half, torch.float, torch.double]
+    complex_types = [torch.complex64, torch.complex128]
+    default_types = [torch.float, torch.double]
+    default_types_cuda = [torch.half, torch.float, torch.double]
+
+    def test_fn(self, device):
+        op = getattr(torch, op_str)
+
+        my_default_types = default_types
+        my_float_types = float_types
+        if self.device_type == 'cuda':
+            my_default_types = default_types_cuda
+            my_float_types = float_types_cuda
+        if self.device_type in unary_ufuncs[op_str].complex_on:
+            my_float_types.extend(complex_types)
+
+        prev_def_type = torch.get_default_dtype()
+        try:
+            for in_type, def_type in product(int_types, my_default_types):
+                torch.set_default_dtype(def_type)
+                t = torch.tensor((1,), device=device, dtype=in_type)
+                self.assertEqual(op(t).dtype, def_type)
+        finally:
+            torch.set_default_dtype(prev_def_type)
+
+        for in_type in my_float_types:
+            t = torch.tensor((1,), device=device, dtype=in_type)
+            self.assertEqual(op(t).dtype, in_type)
+
+        only_cpu_ops = ['atan', 'cos', 'cosh', 'erf', 'erfc', 'exp', 'tan', 'tanh']
```

Comment:
Give this a more descriptive name, like "only_out_on_cpu"

Reply:
Thanks, this sounds reasonable! (https://github.com/pytorch/pytorch/pull/33322/commits/0c9463aabee3a0449cdeae9560b14ab5a7ef5f20).